### PR TITLE
feat: add QueryAsync overload with Type parameter (#230)

### LINKED ISSUE #230: Make QueryAsync accept a Type instead of generic
__Proposal:__
Make QueryAsync accept a `Type` parameter instead of generic so it will work also when we only now the Type on runtime.

__Current behavior:__
It's only whether `QueryAsync` which returns `FluxTable` or `QueryAsync<T>` which returns `T`

__Desired behavior:__
I would like to see a `Type` parameter on the `QueryAsync` method for the mapper.

__Alternatives considered:__
[Reflection](https://stackoverflow.com/a/232621/2770573)

__Use case:__
When we use mappers for `InfluxDbEntity` to e.g. `SensorData` and we use software like `AutoMapper` or create our own Mapper we probably don't know the `Type` of the entity while programming, but only on run time.

### PR DESCRIPTION
**Pull request by maxreb**
Closes #230 

## Proposed Changes

This commit will add an overload to `QueryAsync` to allow a `Type` as parameter for the POCO object.

## Checklist

- [x] CHANGELOG.md updated
- [x] Rebased/mergeable
- [x] A test has been added if appropriate
- [x] `dotnet test` completes successfully
- [x] Commit messages are in [semantic format](https://seesparkbox.com/foundry/semantic_commit_messages)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ### Bug Fixes
 1. [#221](https://github.com/influxdata/influxdb-client-csharp/pull/221): Parsing infinite numbers
 2. [#229](https://github.com/influxdata/influxdb-client-csharp/pull/229): Fix cookie handling in session mode
+3. [#230](https://github.com/influxdata/influxdb-client-csharp/issues/230): Adds an `Type` overload for POCOs to `QueryAsync`
```
Comment (bednar) [inline CHANGELOG.md:6]: Please, use link to Pull Request instead to issue.

```suggestion
3. [#232](https://github.com/influxdata/influxdb-client-csharp/pull/232): Adds an `Type` overload for POCOs to `QueryAsync`
```

File: Client.Core/Flux/Internal/FluxResultMapper.cs [hunk 2]
```diff
@@ -29,164 +29,180 @@
...
+
+			if (value is DateTime dateTime)
+			{
+				return Instant.FromDateTimeUtc(dateTime);
+			}
+
+			throw new InvalidCastException(
+				$"Object value of type {value.GetType().Name} cannot be converted to {nameof(Instant)}");
+		}
+	}
```
Comment (bednar) [inline Client.Core/Flux/Internal/FluxResultMapper.cs:207]: Please use `spaces` for indentation. A tab could be a different number of columns depending on your environment, but a space is always one column and the code will be easier to review.

File: Client.Core/Flux/Internal/IFluxResultMapper.cs [hunk 3]
```diff
@@ -14,5 +15,7 @@ public interface IFluxResultMapper
         /// <typeparam name="T">Type of DomainObject</typeparam>
         /// <returns>Converted DomainObject</returns>
         T ConvertToEntity<T>(FluxRecord fluxRecord);
+
+        object ConvertToEntity(FluxRecord fluxRecord, Type type);
```
Comment (bednar) [inline Client.Core/Flux/Internal/IFluxResultMapper.cs:19]: This is a breaking change. We have to increase major version of the client and add proper description of the change in `CHANGELOG.md`... or avoid this breaking change. 

Please also add documentation comments.

File: Client/Internal/DefaultDomainObjectMapper.cs [hunk 4]
```diff
@@ -1,26 +1,32 @@
...
+		public object ConvertToEntity(FluxRecord fluxRecord, Type type)
+		{
+			return _resultMapper.ToPoco(fluxRecord, type);
+		}
+
+		public PointData ConvertToPointData<T>(T entity, WritePrecision precision)
+		{
+			return _measurementMapper.ToPoint(entity, precision);
+		}
+	}
```
Comment (bednar) [inline Client/Internal/DefaultDomainObjectMapper.cs:31]: Please use `spaces` for indentation. A tab could be a different number of columns depending on your environment, but a space is always one column and the code will be easier to review.

**Review by bednar — CHANGES_REQUESTED** [review]
Thanks for your PR 👍 
There is a few requirements that should be satisfy before we accept the PR:

File: Client.Core/Flux/Internal/IFluxResultMapper.cs (hunk 3 above)
Comment (maxreb) [reply Client.Core/Flux/Internal/IFluxResultMapper.cs:19]: this is a tough one... I could create a new Interface called `IFluxResultMapperOfType` or something and then check in the `?`QueryAsync` method if the `IFluxResultMapper` is inheritable from `IFluxResultMapperOfType` and if not throw a new exception which says something like 
> IFluxResultMapper needs also implementing IFluxResultMapperOfType to call this method
but this feels quite hackish, what do you think?

Comment (bednar) [reply Client.Core/Flux/Internal/IFluxResultMapper.cs:19]: > but this feels quite hackish, what do you think?

Yeah, you are right. Let's increate major version in `*.csproj` to version `3.0.0`.

**Comment by maxreb** [comment]
okay I changed the code according to your comments

File: Client/QueryApi.cs [hunk 5]
```diff
@@ -735,6 +735,116 @@ public Task QueryAsync<T>(Query query, string org, Action<ICancellable, T> onNex
...
+        /// <param name="onComplete">the callback to consume a notification about successfully end of stream</param>
+        /// <param name="pocoType">the type of measurement</param>
+        /// <returns>async task</returns>
+        public Task QueryAsync(string query, string org, Action<ICancellable, object> onNext, Action<Exception> onError,
+            Action onComplete, Type pocoType)
+        {
+            return QueryAsync(CreateQuery(query, DefaultDialect), org, onNext, onError, onComplete, pocoType);
+        }
+
+        public Task QueryAsync(Query query, string org, Action<ICancellable, object> onNext, Action<Exception> onError,
```
Comment (bednar) [inline Client/QueryApi.cs:834]: Please, add documentation comments.

**Review by bednar — CHANGES_REQUESTED** [review]
Thanks again for your PR 👍. We are almost done.  

Please also update the major version in following files:
    - `./Client/Linq/Client.Linq.csproj`
    - `./Client.Legacy/Client.Legacy.csproj`
    - `./Examples/Examples.csproj`

and check last requirement:

**Comment by maxreb** [comment]
@bednar : done

**Review by bednar — APPROVED** [review]
LGTM 👍 

I will merge this PR into master after release `2.1.0` at Friday.

**Pull request merged**